solver/llbsolver/proc/provenance.go

```diff
@@ -0,0 +1,237 @@
+package proc
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	slsa "github.com/in-toto/in-toto-golang/in_toto/slsa_provenance/v0.2"
+	"github.com/moby/buildkit/cache"
+	"github.com/moby/buildkit/cache/config"
+	"github.com/moby/buildkit/exporter/containerimage/exptypes"
+	"github.com/moby/buildkit/frontend"
+	gatewaypb "github.com/moby/buildkit/frontend/gateway/pb"
+	"github.com/moby/buildkit/identity"
+	"github.com/moby/buildkit/solver"
+	"github.com/moby/buildkit/solver/llbsolver"
+	"github.com/moby/buildkit/solver/result"
+	binfotypes "github.com/moby/buildkit/util/buildinfo/types"
+	provenance "github.com/moby/buildkit/util/provenance"
+	"github.com/moby/buildkit/worker"
+	digest "github.com/opencontainers/go-digest"
+	ocispecs "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
+)
+
+var BuildKitBuildType = "https://mobyproject.org/buildkit@v1"
+
+func ProvenanceProcessor(attrs map[string]string) llbsolver.Processor {
+	return func(ctx context.Context, res *frontend.Result, s *llbsolver.Solver, j *solver.Job) (*frontend.Result, error) {
+		if len(res.Refs) == 0 {
+			return nil, errors.New("provided result has no refs")
+		}
+
+		platformsBytes, ok := res.Metadata[exptypes.ExporterPlatformsKey]
+		if !ok {
+			return nil, errors.Errorf("unable to collect multiple refs, missing platforms mapping")
+		}
+
+		var ps exptypes.Platforms
+		if len(platformsBytes) > 0 {
+			if err := json.Unmarshal(platformsBytes, &ps); err != nil {
+				return nil, errors.Wrapf(err, "failed to parse platforms passed to sbom processor")
+			}
+		}
+
+		buildID := identity.NewID()
+
+		var reproducible bool
+		if v, ok := attrs["reproducible"]; ok {
+			b, err := strconv.ParseBool(v)
+			if err != nil {
+				return nil, errors.Wrapf(err, "failed to parse reproducible flag %q", v)
+			}
```

Comment:
Maybe warn if `v` is true, until we can support repro

Reply:
I think this would always be passed by the user. Even if network is disabled, a process can read something from hardware or just run a `for` loop and count how much time it takes. What we can detect automatically is that parameters and materials are complete and there is no network access(hermetic). These are all requirements for SLSA4 while the full byte-by-byte reproducibility is not.